core/mm/core_mmu.c

```diff
@@ -950,6 +950,50 @@ static void check_sec_nsec_mem_config(void)
 	}
 }
 
+#if defined(CFG_CORE_SEL2_SPMC) && defined(CFG_CORE_PHYS_RELOCATABLE)
+static void collect_device_mem_ranges(struct tee_mmap_region *memory_map,
+				      size_t num_elems, size_t *last)
+{
+	const char *dt_device_match_table = {
+		"arm,ffa-manifest-device-regions",
+	};
+	void *fdt = (void *)manifest_dtb;
+	const char *name = NULL;
+	uint64_t pages_cnt = 0;
+	uint64_t base = 0;
+	int subnode = 0;
+	int node = 0;
+
+	node = fdt_node_offset_by_compatible(fdt, 0, dt_device_match_table);
+	if (node < 0)
+		return;
+
+	fdt_for_each_subnode(subnode, fdt, node) {
+		name = fdt_get_name(fdt, subnode, NULL);
+		if (!name)
+			continue;
+
+		if (dt_getprop_as_number(fdt, subnode, "base-address",
+					 &base)) {
+			EMSG("Mandatory field is missing: base-address");
+			panic();
+		}
+
+		if (dt_getprop_as_number(fdt, subnode, "pages-count",
+					 &pages_cnt)) {
+			EMSG("Mandatory field is missing: pages-count");
+			panic();
+		}
+
+		add_phys_mem(memory_map, num_elems, name, MEM_AREA_IO_SEC,
+			     ROUNDDOWN(base, CORE_MMU_PGDIR_SIZE),
```

Comment:
I think we should use 4k boundary, so no rounding is needed.
We should also be able to assume that the base is correctly aligned or we should ignore this entry.

Reply:
How about：
```
add_phys_mem(memory_map, num_elems, name, MEM_AREA_IO_SEC,
	     base, base + pages_cnt * SMALL_PAGE_SIZE , last);
```